File: Python/Deque.py
append(item): Adds an item to the right end of the deque.
Time complexity: Amortized O(1) (constant time on average). Space complexity: O(1).

appendleft(item): Adds an item to the left end of the deque.
Time complexity: Amortized O(1) (constant time on average). Space complexity: O(1).

pop(): Removes and returns the rightmost item from the deque.
Time complexity: O(1). Space complexity: O(1).

popleft(): Removes and returns the leftmost item from the deque.
Time complexity: O(1). Space complexity: O(1).

extend(iterable): Extends the deque by appending elements from the iterable to the right end.
Time complexity: O(k), where k is the length of the iterable. Space complexity: O(k).

extendleft(iterable): Extends the deque by appending elements from the iterable to the left end.
Time complexity: O(k), where k is the length of the iterable. Space complexity: O(k).

rotate(k): Rotates the deque k steps to the right if k is positive, or to the left if k is negative.
Time complexity: O(k), where k is the number of steps to rotate. Space complexity: O(1).

clear(): Removes all elements from the deque, leaving it empty.
Time complexity: O(1). Space complexity: O(1).

count(item): Returns the number of occurrences of the item in the deque.
Time complexity: O(n), where n is the number of elements in the deque. Space complexity: O(1).

index(item, start, end): Returns the index of the first occurrence of the item in the deque between the start and end indices.
Time complexity: O(n), where n is the number of elements in the deque. Space complexity: O(1).

remove(item): Removes the first occurrence of the item from the deque.
Time complexity: O(n), where n is the number of elements in the deque. Space complexity: O(1).

Deque implementation using list:

class Deque:
    def __init__(self):
        self.data = []

    def append(self, item):
        self.data.append(item)

    def appendleft(self, item):
        self.data.insert(0, item)

    def pop(self):
        if not self.is_empty():
            return self.data.pop()
        else:
            raise IndexError("Deque is empty")

    def popleft(self):
        if not self.is_empty():
            return self.data.pop(0)
        else:
            raise IndexError("Deque is empty")

    def extend(self, iterable):
        self.data.extend(iterable)

    def extendleft(self, iterable):
        for item in iterable:
            self.appendleft(item)

    def rotate(self, k):
        self.data = self.data[-k:] + self.data[:-k]

    def clear(self):
        self.data.clear()

    def count(self, item):
        return self.data.count(item)

    def index(self, item, start=None, end=None):
        return self.data.index(item, start, end)

    def remove(self, item):
        self.data.remove(item)

    def is_empty(self):
        return len(self.data) == 0

    def size(self):
        return len(self.data)

Dynamic size: Doubly linked lists allow for efficient resizing and dynamic memory allocation as
elements are added or removed from the deque.

Constant-time insertion and removal: Insertion and removal operations at both ends of a
doubly linked list can be performed in constant time, while arrays/vectors require shifting/moving
elements.

Efficient element removal: Removing elements from a doubly linked list requires updating neighboring
pointers, resulting in more efficient removal compared to arrays/vectors.

Easy reordering of elements: Doubly linked lists allow for easy reordering of elements by changing
pointers, making them useful in scenarios requiring frequent element rearrangement.

Reduced memory overhead: Doubly linked lists offer more efficient memory utilization as they don't
require upfront allocation of a fixed-size buffer like arrays/vectors.

Implementation using Doubly Linked List:

class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None

class Deque:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def append(self, item):
        new_node = Node(item)
        if self.is_empty():
            self.head = new_node
            self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
        self.size += 1

    def appendleft(self, item):
        new_node = Node(item)
        if self.is_empty():
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        self.size += 1

    def pop(self):
        if self.is_empty():
            raise IndexError("Deque is empty")
        item = self.tail.value
        if self.size == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
        self.size -= 1
        return item

    def popleft(self):
        if self.is_empty():
            raise IndexError("Deque is empty")
        item = self.head.value
        if self.size == 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None
        self.size -= 1
        return item

    def is_empty(self):
        return self.size == 0

    def get_size(self):
        return self.size

File: C++/Deque.cpp
push_back(item): Adds an item to the rear of the deque.
Time complexity: Amortized O(1) (constant time on average). Space complexity: O(1).

push_front(item): Adds an item to the front of the deque.
Time complexity: Amortized O(1) (constant time on average). Space complexity: O(1).

pop_back(): Removes the item from the rear of the deque.
Time complexity: Amortized O(1) (constant time on average). Space complexity: O(1).

pop_front(): Removes the item from the front of the deque.
Time complexity: Amortized O(1) (constant time on average). Space complexity: O(1).

front(): Returns a reference to the item at the front of the deque.
Time complexity: O(1). Space complexity: O(1).

back(): Returns a reference to the item at the rear of the deque.
Time complexity: O(1). Space complexity: O(1).

at(index): Returns a reference to the item at the specified index in the deque.
Time complexity: O(1). Space complexity: O(1).

empty(): Checks if the deque is empty.
Time complexity: O(1). Space complexity: O(1).

size(): Returns the number of elements in the deque.
Time complexity: O(1). Space complexity: O(1).

clear(): Removes all elements from the deque, leaving it with a size of 0.
Time complexity: Linear O(N) in the number of elements. Space complexity: O(1).


#include <iostream>
#include <stdexcept>

template<typename T>
class Deque {
private:
    T* buffer;
    int capacity;
    int frontIndex;
    int rearIndex;
    int size;

public:
    Deque() : capacity(8), frontIndex(0), rearIndex(0), size(0) {
        buffer = new T[capacity];
    }

    ~Deque() {
        delete[] buffer;
    }

    void push_back(const T& item) {
        if (size == capacity) {
            resize(capacity * 2);
        }
        buffer[rearIndex] = item;
        rearIndex = (rearIndex + 1) % capacity;
        size++;
    }

    void push_front(const T& item) {
        if (size == capacity) {
            resize(capacity * 2);
        }
        frontIndex = (frontIndex - 1 + capacity) % capacity;
        buffer[frontIndex] = item;
        size++;
    }

    T pop_back() {
        if (size == 0) {
            throw std::runtime_error("Deque is empty");
        }
        rearIndex = (rearIndex - 1 + capacity) % capacity;
        T item = buffer[rearIndex];
        size--;
        if (size <= capacity / 4) {
            resize(capacity / 2);
        }
        return item;
    }

    T pop_front() {
        if (size == 0) {
            throw std::runtime_error("Deque is empty");
        }
        T item = buffer[frontIndex];
        frontIndex = (frontIndex + 1) % capacity;
        size--;
        if (size <= capacity / 4) {
            resize(capacity / 2);
        }
        return item;
    }

    T& front() {
        if (size == 0) {
            throw std::runtime_error("Deque is empty");
        }
        return buffer[frontIndex];
    }

    T& back() {
        if (size == 0) {
            throw std::runtime_error("Deque is empty");
        }
        return buffer[(rearIndex - 1 + capacity) % capacity];
    }

    bool empty() {
        return size == 0;
    }

    int get_size() {
        return size;
    }

private:
    void resize(int newCapacity) {
        T* newBuffer = new T[newCapacity];
        for (int i = 0; i < size; i++) {
            newBuffer[i] = buffer[(frontIndex + i) % capacity];
        }
        delete[] buffer;
        buffer = newBuffer;
        capacity = newCapacity;
        frontIndex = 0;
        rearIndex = size;
    }
};


Using a std::vector for implementing a deque-like data structure is indeed a valid approach.
In fact, the C++ Standard Library (std::deque) itself uses a combination of multiple std::vector segments to implement a deque.

The reason for not using std::vector directly in the provided implementation is to demonstrate an alternative approach that
uses a dynamically allocated buffer as the underlying storage for the deque.
This approach highlights the core concepts of a deque, such as the ability to efficiently insert and remove elements at both ends,
without relying on the more complex structure of a segmented vector.

However, using std::vector as the underlying container for implementing a deque has its advantages.

Some benefits of using std::vector include:

Simplicity: std::vector provides a straightforward and intuitive interface, making the implementation and
usage of a deque-like data structure more concise.
Continuous memory: std::vector guarantees that its elements are stored in contiguous memory, which can improve cache performance
in certain scenarios.
Efficiency: std::vector provides efficient random access and supports operations like insert() and erase() in the middle,
which can be useful in some deque operations.
If you choose to use std::vector for implementing a deque, you can utilize its existing functions like push_back(),
push_front(), pop_back(), pop_front(), and other operations to achieve the desired functionality.

Implemntation using vector


#include <iostream>
#include <vector>
#include <stdexcept>

template<typename T>
class Deque {
private:
    std::vector<T> data;

public:
    void push_back(const T& item) {
        data.push_back(item);
    }

    void push_front(const T& item) {
        data.insert(data.begin(), item);
    }

    T pop_back() {
        if (is_empty()) {
            throw std::runtime_error("Deque is empty");
        }
        T item = data.back();
        data.pop_back();
        return item;
    }

    T pop_front() {
        if (is_empty()) {
            throw std::runtime_error("Deque is empty");
        }
        T item = data.front();
        data.erase(data.begin());
        return item;
    }

    T& front() {
        if (is_empty()) {
            throw std::runtime_error("Deque is empty");
        }
        return data.front();
    }

    T& back() {
        if (is_empty()) {
            throw std::runtime_error("Deque is empty");
        }
        return data.back();
    }

    bool is_empty() {
        return data.empty();
    }

    int size() {
        return data.size();
    }
};

Dynamic size: Doubly linked lists allow for efficient resizing and dynamic memory allocation as
elements are added or removed from the deque.

Constant-time insertion and removal: Insertion and removal operations at both ends of a
doubly linked list can be performed in constant time, while arrays/vectors require shifting/moving
elements.

Efficient element removal: Removing elements from a doubly linked list requires updating neighboring
pointers, resulting in more efficient removal compared to arrays/vectors.

Easy reordering of elements: Doubly linked lists allow for easy reordering of elements by changing
pointers, making them useful in scenarios requiring frequent element rearrangement.

Reduced memory overhead: Doubly linked lists offer more efficient memory utilization as they don't
require upfront allocation of a fixed-size buffer like arrays/vectors.

Implemntation using Doubly Linked List

#include <iostream>
#include <stdexcept>

template<typename T>
struct Node {
    T value;
    Node* prev;
    Node* next;

    Node(const T& val) : value(val), prev(nullptr), next(nullptr) {}
};

template<typename T>
class Deque {
private:
    Node<T>* head;
    Node<T>* tail;
    int size;

public:
    Deque() : head(nullptr), tail(nullptr), size(0) {}

    ~Deque() {
        clear();
    }

    void append(const T& item) {
        Node<T>* new_node = new Node<T>(item);
        if (is_empty()) {
            head = new_node;
            tail = new_node;
        } else {
            new_node->prev = tail;
            tail->next = new_node;
            tail = new_node;
        }
        size++;
    }

    void appendleft(const T& item) {
        Node<T>* new_node = new Node<T>(item);
        if (is_empty()) {
            head = new_node;
            tail = new_node;
        } else {
            new_node->next = head;
            head->prev = new_node;
            head = new_node;
        }
        size++;
    }

    T pop() {
        if (is_empty()) {
            throw std::runtime_error("Deque is empty");
        }
        T item = tail->value;
        Node<T>* temp = tail;
        tail = tail->prev;
        if (tail) {
            tail->next = nullptr;
        } else {
            head = nullptr;
        }
        delete temp;
        size--;
        return item;
    }

    T popleft() {
        if (is_empty()) {
            throw std::runtime_error("Deque is empty");
        }
        T item = head->value;
        Node<T>* temp = head;
        head = head->next;
        if (head) {
            head->prev = nullptr;
        } else {
            tail = nullptr;
        }
        delete temp;
        size--;
        return item;
    }

    bool is_empty() {
        return size == 0;
    }

    int get_size() {
        return size;
    }

    void clear() {
        Node<T>* current = head;
        while (current) {
            Node<T>* next = current->next;
            delete current;
            current = next;
        }
        head = nullptr;
        tail = nullptr;
        size = 0;
    }
};
